subsys/net/ip/connection.c

```diff
@@ -576,94 +576,283 @@ static bool conn_are_endpoints_valid(struct net_pkt *pkt, uint8_t family,
+				       uint8_t proto)
 {
-	struct net_pkt *raw_pkt;
-	struct net_pkt_cursor cur;
+	uint8_t pkt_family = net_pkt_family(pkt);
+	struct net_conn *conn;
 
-	net_pkt_cursor_backup(pkt, &cur);
-	net_pkt_cursor_init(pkt);
+	if (pkt_family != AF_INET && pkt_family != AF_INET6) {
+		return NET_CONTINUE;
+	}
 
-	NET_DBG("[%p] raw IP match found cb %p ud %p", conn, conn->cb,
-		conn->user_data);
+	NET_DBG("Check %s listener for pkt %p family %d",
+		net_proto2str(net_pkt_family(pkt), proto), pkt,
+		net_pkt_family(pkt));
 
-	raw_pkt = net_pkt_clone(pkt, K_MSEC(CONFIG_NET_CONN_PACKET_CLONE_TIMEOUT));
-	if (raw_pkt == NULL) {
-		goto out;
-	}
+	k_mutex_lock(&conn_lock, K_FOREVER);
 
-	if (conn->cb(conn, raw_pkt, NULL, NULL, conn->user_data) == NET_DROP) {
-		net_pkt_unref(raw_pkt);
+	SYS_SLIST_FOR_EACH_CONTAINER(&conn_used, conn, node) {
+		if (!is_iface_matching(conn, pkt)) {
+			continue; /* wrong interface */
+		}
+
+		if (conn->family != pkt_family) {
+			continue; /* wrong family */
+		}
+
+		if (conn->type != SOCK_RAW) {
+			continue; /* wrong type */
+		}
+
+		if (conn->proto != proto && conn->proto != IPPROTO_IP) {
+			continue; /* wrong protocol */
+		}
+
+		/* Apply protocol-specific matching criteria... */
+
+		if ((conn->flags & NET_CONN_LOCAL_ADDR_SET) &&
+		    !conn_addr_cmp(pkt, ip_hdr, &conn->local_addr, false)) {
+			continue; /* wrong local address */
+		}
+
+		conn_raw_socket_deliver(pkt, conn, true);
 	}
 
-out:
-	net_pkt_cursor_restore(pkt, &cur);
+	k_mutex_unlock(&conn_lock);
+
+	/* Raw IP packets are passed further in the stack regardless. */
+	return NET_CONTINUE;
 }
 #else
-static void conn_raw_ip_socket(struct net_pkt *pkt, struct net_conn *conn)
+enum net_verdict net_conn_raw_ip_input(struct net_pkt *pkt,
+				       union net_ip_header *ip_hdr,
+				       uint8_t proto)
 {
 	ARG_UNUSED(pkt);
-	ARG_UNUSED(conn);
+	ARG_UNUSED(ip_hdr);
+	ARG_UNUSED(proto);
+
+	return NET_CONTINUE;
 }
 #endif /* defined(CONFIG_NET_SOCKETS_INET_RAW) */
 
+#if defined(CONFIG_NET_SOCKETS_CAN)
+enum net_verdict net_conn_can_input(struct net_pkt *pkt, uint8_t proto)
+{
+	struct net_conn *best_match = NULL;
+	struct net_conn *conn;
+	net_conn_cb_t cb = NULL;
+	void *user_data = NULL;
+
+	/* Only accept input with AF_CAN family and CAN_RAW protocol. */
+	if (net_pkt_family(pkt) != AF_CAN || proto != CAN_RAW) {
+		return NET_DROP;
+	}
+
+	NET_DBG("Check %s listener for pkt %p family %d",
+		net_proto2str(net_pkt_family(pkt), proto), pkt,
+		net_pkt_family(pkt));
+
+	k_mutex_lock(&conn_lock, K_FOREVER);
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&conn_used, conn, node) {
+		if (!is_iface_matching(conn, pkt)) {
+			continue; /* wrong interface */
+		}
+
+		if (conn->family != AF_CAN) {
+			continue; /* wrong family */
+		}
+
+		if (conn->proto != CAN_RAW) {
+			continue; /* wrong protocol */
+		}
+
+		best_match = conn;
+	}
+
+	if (best_match) {
+		cb = best_match->cb;
+		user_data = best_match->user_data;
+	}
+
+	k_mutex_unlock(&conn_lock);
+
+	if (cb) {
```

Comment:
Nit
```suggestion
	if (best_match != NULL) {
		cb = best_match->cb;
		user_data = best_match->user_data;
	}

	k_mutex_unlock(&conn_lock);

	if (cb != NULL) {
```

Reply:
Fixed here and similar case in `net_conn_input()`